### PR TITLE
Apply dev specs for dependencies of roots

### PR DESCRIPTION
**Pull request by becker33**
Fixes a bug on develop and 0.18.0.

Currently, develop specs that are not roots and are not explicitly listed dependencies of the roots are not applied.

### CODE REVIEW

**Review by tgamblin — DISMISSED** [review]

**Review by tgamblin — DISMISSED** [review]

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -1841,11 +1849,23 @@ def setup(self, driver, specs, reuse=None):
...
+        # they will be used in addition to command line specs
+        # in determining known versions/targets/os
+        env = ev.active_environment()
+        dev_specs = tuple(
+            spack.spec.Spec(info['spec']).constrained(
+                'dev_path=%s' % info['path']
+            )
+            for name, info in env.dev_specs.items()
+        ) if env else tuple()
+        specs = tuple(specs)  # ensure compatible types to add
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py:1964]: There is some context in which these come in as tuples and come in which they come in as lists, converting to tuples ensures we're able to add the two types together and in this context an immutable type makes sense.

**Comment by becker33** [comment]
@spackbot fix style

**Review by tgamblin — DISMISSED** [review]

File: lib/spack/spack/solver/asp.py [hunk 2]
```diff
@@ -1546,8 +1556,13 @@ def key_fn(item):
                 # if there is a concrete version on the CLI *that we know nothing
                 # about*, add it to the known versions. Use idx=0, which is the
                 # best possible, so they're guaranteed to be used preferentially.
+                if i < len(specs):
+                    origin = version_provenance.spec
+                else:
+                    origin = version_provenance.dev_spec
+
```
Comment (tgamblin) [inline lib/spack/spack/solver/asp.py]: This is hard to read and it's hard to understand that you're basically doing the same thing to two lists with one tweak.

This should be refactored by pulling the loop nest out as its own function like `_add_concrete_versions(self, specs, origin)`, then refactor this block as:

```python
self._add_concrete_versions(specs, version_provenance.spec)
self._add_concrete_versions(dev_specs, version_provenance.dev_spec)
```

File: lib/spack/spack/solver/asp.py [hunk 3]
```diff
@@ -1524,7 +1532,9 @@ def key_fn(item):
                     DeclaredVersion(version=ver, idx=idx, origin=version_provenance.packages_yaml)
                 )
 
-        for spec in specs:
+        # enumerate so we can determine which list it came from
+        # cast to list to ensure addable types
```
Comment (tgamblin) [inline lib/spack/spack/solver/asp.py]: I don't understand the "cast to list" comment (also -- Python has no casting? I can google python and casting and get some results, but in Python you're really just constructing a list out of a set or whatever.  I'd avoid the word here)

File: lib/spack/spack/solver/asp.py (hunk 1 above)
Comment (tgamblin) [reply lib/spack/spack/solver/asp.py:1964]: oh I think the comment above was meant for here.

File: lib/spack/spack/solver/asp.py [hunk 4]
```diff
@@ -1938,11 +1953,28 @@ def setup(self, driver, specs, reuse=None):
...
+            else tuple()
+        )
+        specs = tuple(specs)  # ensure compatible types to add
+
         # get possible compilers
         self.possible_compilers = self.generate_possible_compilers(specs)
 
         # traverse all specs and packages to build dict of possible versions
-        self.build_version_dict(possible, specs)
+        self.build_version_dict(possible, specs, dev_specs)
```
Comment (tgamblin) [inline lib/spack/spack/solver/asp.py]: to make this more clear I would factor out the bottom half of `build_version_dict` as I mentioned above, but call it here:

```python
    self.build_version_dict(possible)
    self.add_concrete_versions_from_specs(specs, version_provenance.spec)
    self.add_concrete_versions_from_specs(dev_specs, version_provenance.dev_spec)
```

That function is doing too much stuff -- the two halves are really two functions and conflating the parameters here makes the code hard to follow.

**Review by tgamblin — CHANGES_REQUESTED** [review]

**Comment by tgamblin** [comment]
@becker33 I went ahead and did the changes I suggested -- see if you like them.

**Comment by becker33** [comment]
Yes LGTM, thanks!

**Review by tgamblin — APPROVED** [review]
